File: app/src/main/java/com/example/englishhindi/repository/QuizRepository.java
package com.example.englishhindi.repository;

import android.app.Application;
import android.os.AsyncTask;

import androidx.lifecycle.LiveData;

import com.example.englishhindi.database.QuizDao;
import com.example.englishhindi.database.WordDatabase;
import com.example.englishhindi.model.Quiz;

import java.util.List;

public class QuizRepository {
    
    private QuizDao quizDao;
    
    public QuizRepository(Application application) {
        WordDatabase database = WordDatabase.getInstance(application);
        quizDao = database.quizDao();
    }
    
    // Insert a new quiz
    public void insert(Quiz quiz) {
        new InsertQuizAsyncTask(quizDao).execute(quiz);
    }
    
    // Update an existing quiz
    public void update(Quiz quiz) {
        new UpdateQuizAsyncTask(quizDao).execute(quiz);
    }
    
    // Delete a quiz
    public void delete(Quiz quiz) {
        new DeleteQuizAsyncTask(quizDao).execute(quiz);
    }
    
    // Get quizzes for a specific lesson
    public LiveData<List<Quiz>> getQuizzesByLesson(int lessonId) {
        return quizDao.getQuizzesByLesson(lessonId);
    }
    
    // Get random quizzes
    public LiveData<List<Quiz>> getRandomQuizzes(int limit) {
        return quizDao.getRandomQuizzes(limit);
    }
    
    // Get quizzes by difficulty
    public LiveData<List<Quiz>> getQuizzesByDifficulty(int maxDifficulty, int limit) {
        return quizDao.getQuizzesByDifficulty(maxDifficulty, limit);
    }
    
    // Get a specific quiz by ID
    public LiveData<Quiz> getQuizById(int quizId) {
        return quizDao.getQuizById(quizId);
    }
    
    // Get quiz count for a lesson
    public int getQuizCountForLesson(int lessonId) {
        return quizDao.getQuizCountForLesson(lessonId);
    }
    
    // Get quizzes by type
    public LiveData<List<Quiz>> getQuizzesByType(String quizType, int limit) {
        return quizDao.getQuizzesByType(quizType, limit);
    }
    
    // Insert quiz with details
    public void insertQuizDetails(int lessonId, String question, String questionHindi, 
                                 String correctAnswer, String wrongAnswer1, 
                                 String wrongAnswer2, String wrongAnswer3,
                                 String explanation, String explanationHindi, 
                                 String type, int difficulty) {
        new InsertQuizDetailsAsyncTask(quizDao, lessonId, question, questionHindi,
                                       correctAnswer, wrongAnswer1, wrongAnswer2, wrongAnswer3,
                                       explanation, explanationHindi, type, difficulty).execute();
    }
    
    // AsyncTask classes for database operations
    private static class InsertQuizAsyncTask extends AsyncTask<Quiz, Void, Void> {
        private QuizDao quizDao;
        
        private InsertQuizAsyncTask(QuizDao quizDao) {
            this.quizDao = quizDao;
        }
        
        @Override
        protected Void doInBackground(Quiz... quizzes) {
            quizDao.insert(quizzes[0]);
            return null;
        }
    }
    
    private static class UpdateQuizAsyncTask extends AsyncTask<Quiz, Void, Void> {
        private QuizDao quizDao;
        
        private UpdateQuizAsyncTask(QuizDao quizDao) {
            this.quizDao = quizDao;
        }
        
        @Override
        protected Void doInBackground(Quiz... quizzes) {
            quizDao.update(quizzes[0]);
            return null;
        }
    }
    
    private static class DeleteQuizAsyncTask extends AsyncTask<Quiz, Void, Void> {
        private QuizDao quizDao;
        
        private DeleteQuizAsyncTask(QuizDao quizDao) {
            this.quizDao = quizDao;
        }
        
        @Override
        protected Void doInBackground(Quiz... quizzes) {
            quizDao.delete(quizzes[0]);
            return null;
        }
    }
    
    private static class InsertQuizDetailsAsyncTask extends AsyncTask<Void, Void, Void> {
        private QuizDao quizDao;
        private int lessonId;
        private String question;
        private String questionHindi;
        private String correctAnswer;
        private String wrongAnswer1;
        private String wrongAnswer2;
        private String wrongAnswer3;
        private String explanation;
        private String explanationHindi;
        private String type;
        private int difficulty;
        
        private InsertQuizDetailsAsyncTask(QuizDao quizDao, int lessonId, String question, 
                                          String questionHindi, String correctAnswer, 
                                          String wrongAnswer1, String wrongAnswer2, 
                                          String wrongAnswer3, String explanation, 
                                          String explanationHindi, String type, int difficulty) {
            this.quizDao = quizDao;
            this.lessonId = lessonId;
            this.question = question;
            this.questionHindi = questionHindi;
            this.correctAnswer = correctAnswer;
            this.wrongAnswer1 = wrongAnswer1;
            this.wrongAnswer2 = wrongAnswer2;
            this.wrongAnswer3 = wrongAnswer3;
            this.explanation = explanation;
            this.explanationHindi = explanationHindi;
            this.type = type;
            this.difficulty = difficulty;
        }
        
        @Override
        protected Void doInBackground(Void... voids) {
            quizDao.insertQuizDetails(lessonId, question, questionHindi, correctAnswer,
                                     wrongAnswer1, wrongAnswer2, wrongAnswer3,
                                     explanation, explanationHindi, type, difficulty);
            return null;
        }
    }
}

File: app/src/main/java/com/example/englishhindi/data/repository/QuizRepository.kt
package com.bhashasetu.app.data.repository

import androidx.lifecycle.LiveData
import com.bhashasetu.app.data.dao.QuizDao
import com.bhashasetu.app.data.model.Quiz
import com.bhashasetu.app.data.model.QuizQuestion
import com.bhashasetu.app.data.relation.QuizWithQuestions
import kotlinx.coroutines.flow.Flow
import java.util.Date

class QuizRepository(private val quizDao: QuizDao) {

    val allQuizzes: Flow<List<Quiz>> = quizDao.getAllQuizzes()
    
    suspend fun insert(quiz: Quiz): Long {
        return quizDao.insert(quiz)
    }
    
    suspend fun update(quiz: Quiz) {
        quizDao.update(quiz)
    }
    
    suspend fun delete(quiz: Quiz) {
        quizDao.delete(quiz)
    }
    
    fun getQuizById(id: Long): Flow<Quiz> {
        return quizDao.getQuizById(id)
    }
    
    fun getQuizzesByType(quizType: String): Flow<List<Quiz>> {
        return quizDao.getQuizzesByType(quizType)
    }
    
    fun getQuizzesByDifficulty(difficulty: Int): Flow<List<Quiz>> {
        return quizDao.getQuizzesByDifficulty(difficulty)
    }
    
    fun getQuizWithQuestions(quizId: Long): Flow<QuizWithQuestions> {
        return quizDao.getQuizWithQuestions(quizId)
    }
    
    fun getAllQuizzesWithQuestions(): Flow<List<QuizWithQuestions>> {
        return quizDao.getAllQuizzesWithQuestions()
    }
    
    suspend fun insertQuestion(question: QuizQuestion) {
        quizDao.insertQuestion(question)
    }
    
    suspend fun updateQuestion(question: QuizQuestion) {
        quizDao.updateQuestion(question)
    }
    
    suspend fun deleteQuestion(question: QuizQuestion) {
        quizDao.deleteQuestion(question)
    }
    
    suspend fun updateLastAttemptAndScore(quizId: Long, lastAttemptDate: Date, score: Int) {
        quizDao.updateLastAttemptAndScore(quizId, lastAttemptDate, score)
    }
    
    suspend fun incrementCompletionCount(quizId: Long) {
        quizDao.incrementCompletionCount(quizId)
    }
    
    fun getQuizzesByLessonId(lessonId: Long): Flow<List<Quiz>> {
        return quizDao.getQuizzesByLessonId(lessonId)
    }
    
    fun getQuizzesByCategoryId(categoryId: Long): Flow<List<Quiz>> {
        return quizDao.getQuizzesByCategoryId(categoryId)
    }
    
    fun getQuestionsByQuizId(quizId: Long): Flow<List<QuizQuestion>> {
        return quizDao.getQuestionsByQuizId(quizId)
    }
    
    fun getCompletedQuizCount(): Flow<Int> {
        return quizDao.getCompletedQuizCount()
    }
    
    fun getAverageQuizScore(): Flow<Float> {
        return quizDao.getAverageQuizScore()
    }
}